File: 剑指offer/52.两个链表的第一个公共节点.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def getIntersectionNode(self, headA: ListNode,
                            headB: ListNode) -> ListNode:
        p = headA
        q = headB

        while p != q:
            if p:
                p = p.next
            else:
                p = headB

            if q:
                q = q.next
            else:
                q = headA

        return p

File: 剑指offer/52.两个链表的第一个公共节点.cpp
#include <iostream>
using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x)
        : val(x), next(NULL)
    {
    }
};

class Solution
{
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB)
    {
        ListNode *p = headA;
        ListNode *q = headB;

        while (p != q)
        {
            if (p != NULL)
            {
                p = p->next;
            }
            else
            {
                p = headB;
            }

            if (q != NULL)
            {
                q = q->next;
            }
            else
            {
                q = headA;
            }
        }

        return p;
    }
};